design/open/uncomitted-gc.md

```diff
@@ -0,0 +1,162 @@
+
+#### S3 Gateway CopyObject
+
+1. Copy object in the same branch will work the same - creating a new staging entry using the existing entry information.
+2. For objects that are not part of the branch, use the underlying adapter copy operation.
+
+#### Move/RenameObject
+Clients working through the S3 Gateway can use the CopyObject + DeleteObject to perform a Rename or Move operation.
+For clients using the OpenAPI this could have been done using StageObject + DeleteObject.
+To continue support of this operation, introduce a new API to rename an object which will be scoped to a single branch.
+
+#### PrepareUncommittedForGC
+
+A new API which will create meta-ranges and ranges for a given branch using its uncommitted data. These files
+will be saved in a designated path used by the GC client to list branch's uncommitted objects.  
+For the purpose of this document we'll call this the `BranchUncommittedMetarange`
+
+### GC Flows
+
+The following describe the GC process run flows on a repository:
+
+#### Flow 1: Clean Run
+
+1. Listing namespace objects
+   1. Read all objects directly from object store (can be done in parallel using the partitions) -> `Store DF`
+   2. Skip partitions that are newer than < TOKEN_EXPIRY_TIME >
+2. Listing of lakeFS repository uncommitted objects
+   1. Mark uncommitted data
+      1. List branches
+      2. Run _PrepareUncommittedForGC_ on all branches
+   2. Get all uncommitted data addresses
+      1. For each branch
+         1. Read all addresses from branch `BranchUncommittedMetarange` -> `Uncommitted DF`
+    >**Note:** To avoid possible bug, uncommitted listing must happen before listing of committed data
+3. Listing of lakeFS repository committed objects
+   1. Get all committed data addresses
+      1. Read all addresses from Repository commits -> `Committed DF`
+4. Find candidates for deletion
+   1. Subtract committed data from all objects (`Store DF` - `Committed DF`)
+   2. Subtract uncommitted data from all objects (`Store DF` - `Uncommitted DF`)
+   3. Filter files in special paths
+   4. The remainder is a list of files which can be safely removed
+5. Save run information under the GC path
+    1. The current run's `Uncommitted DF` (as a parquet file)
+    2. The last read partition
+    3. The GC run start timestamp
+
+#### Flow 2: Optimized Run
+
+Optimized run uses the previous GC run output, to perform a partial scan of the branch to remove uncommitted garbage.
+
+##### Step 1. Analyze Data and Perform Cleanup for old entries (GC client)
+
+1. Read previous run's information
+   1. Previous `Uncommitted DF`
+   2. Last read partition
+   3. Last run's timestamp
+2. Listing of lakeFS repository uncommitted objects  
+    See previous for steps
+3. Listing of lakeFS repository committed objects (optimized)
+   1. Read addresses from branch's new commits (all new commits down to the last GC run timestamp) -> `Committed DF`
```

Comment:
It can be problematic to rely on the timestamp of the commit, because:
1. There can be clock differences.
2. There is a plumbing command that allows to manually set the commit date.

If we go down that road, you should take a margin of safety, and include commits from some time before the last GC run - to avoid race conditions.

Reply:
We're already taking a margin on the GetPhysicalAddress token expiration time - for certainty we can say that GC will not process partitions created in the last 24 hours (though I prefer not stating specific numbers in the proposal). WDYT?

Regarding the plumbing command - can you tell me what it is used for?

Reply:
The plumbing command is used today for testing Garbage Collection.